src/simulators/statevector/statevector_state.hpp

```diff
@@ -634,6 +655,84 @@ void State<statevec_t>::snapshot_matrix_expval(const Operations::Op &op,
   BaseState::qreg_.revert(false);
 }
 
+template <class statevec_t>
+void State<statevec_t>::snapshot_density_matrix(const Operations::Op &op,
+                                                ExperimentData &data,
+                                                SnapshotDataType type) {
+  cmatrix_t reduced_state;
+
+  // Check if tracing over all qubits
+  if (op.qubits.empty()) {
+    reduced_state = cmatrix_t(1, 1);
+    reduced_state[0] = BaseState::qreg_.norm();
+  } else {
+    reduced_state = std::move(density_matrix(op.qubits));
+  }
+
+  // Add density matrix to result data
+  switch (type) {
+    case SnapshotDataType::average:
+      data.add_average_snapshot("density_matrix", op.string_params[0],
+                            BaseState::creg_.memory_hex(), std::move(reduced_state), false);
+      break;
+    case SnapshotDataType::average_var:
+      data.add_average_snapshot("density_matrix", op.string_params[0],
+                            BaseState::creg_.memory_hex(), std::move(reduced_state), true);
+      break;
+    case SnapshotDataType::pershot:
+      data.add_pershot_snapshot("density_matrix", op.string_params[0], std::move(reduced_state));
+      break;
+  }
+}
+
+// Default value that copies using QubitVector.vector method.
+// This is required for Thrust backends which must copy the vector
+// from device memory to host memory
+template <class statevec_t>
+cmatrix_t State<statevec_t>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, std::move(BaseState::qreg_.vector()));
+}
+
+// Overloads for CPU version that can access array data directly
+// without requiring a temporary copy
+template <>
+cmatrix_t State<QV::QubitVector<double>>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.data());
+}
+template <>
+cmatrix_t State<QV::QubitVector<float>>::density_matrix(const reg_t &qubits) {
+  return vec2density(qubits, BaseState::qreg_.data());
+}
+
+template <class statevec_t>
+template <class T>
+cmatrix_t State<statevec_t>::vec2density(const reg_t &qubits, T &&vec) {
```

Comment:
Ok, I see why you are using rvalue reference: Because of the `vector()` method that transforms a pointer to a vector into a vector... by copying everything. If we find out that using `data()` instead of `vector()` is an option, we have to change the signature to get a reference instead.

Reply:
For CPU case it uses `data`, for Thrust/GPU case it uses `vector`, which is why I passed by rvalue